### PR TITLE
Allowing VariantQuery for Feed Generation

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Documentation Update

## Rollback Strategy

Once we merge this code, the new VariantQuery will still not be part of the live production code path.  To do that, we’ll need to enable the `:feed_uses_variant_query_feature` FeatureFlag.

To rollback the change, without a new deploy, we would disable the `:feed_uses_variant_query_feature` FeatureFlag.

## Description

Here's a Loom of the [Code Walkthrough](https://www.loom.com/share/31f06224b61c4f7ca01c85e1fe0c239a).

Apologies for the breadth of this pull request, I had considered many
small commits, but felt that would've been more effort for the value
provided.

This commit includes the following:

- Documentation updates to the feed variant (though not the final pass)
- Renaming and adding RelevancyLevers that help differentiate
- Adding RelevancyLever#range to provide documentation
- Reducing redundent controller logic by making a
  `Articles::Feeds.feed_for` method.
- Adding some configuration validation for RelevancyLevers
- Adding constants for better clarification
- Testing unhappy paths for feed configuration
- Adjusting the module namespace of some objects
- Exposing top-level configurations for variants (along with their
  defaults)
- Creating the VariantyQuery that at present inherits from the
  `Articles::Feeds::WeightedQueryStrategy`

The [./config/feed-variants/README.md](https://github.com/forem/forem/blob/3a9d9ad2d1867589a2d1f2260dab9ac285bd73e6/config/feed-variants/README.md) provides the best starting point.  Also, consider the prior forem/forem#17314 PR (with it's walk through).

As implemented, we can deploy this code to production without using the
new VariantQuery.  Once we toggle on the
`:feed_uses_variant_query_feature` FeatureFlag, it will switch to using
the VariantQuery.  The VariantQuery's two variants and the
internal configuration of `Articles::Feeds::WeightedQueryStrategy`
produce the same query.

The goal of this factor is to allow for a quick on and off toggle of the
feed query; to ensure that what we introduce remains performant.

In addition, I will be recording a code-walkthrough and linking that
recording to the pull request.

### Conceptual Sequence Diagram

The following conceptual sequence diagram illustrates roughly how this new feature behaves; I have not added the "is the feature flag on" bit of logic.

A sequence diagram with the Plantuml Description included below

<details>
<summary>PlantUML syntax</summary>

```plantuml
@startuml
!theme amiga

participant "GET relevancy feed" as GetRequest
participant "AbExperiment.get" as AbExperiment
participant "Articles::Feeds::\nVariantQuery.build_for" as VariantQuery
participant "Articles::Feeds::\nVariantAssembler.call" as Assembler
participant "/config/feed-variants/*.json" as VariantConfig
participant "Articles::Feeds\n.lever_catalog" as LeverCatalog

GetRequest --> AbExperiment : with :user
GetRequest <-- AbExperiment : :variant
GetRequest --> VariantQuery : with :user, :variant
VariantQuery --> Assembler : with :variant
Assembler --> VariantConfig : with :variant
Assembler <-- VariantConfig : :variant_config
Assembler --> LeverCatalog : with :variant_config
Assembler <-- LeverCatalog : :query_config
VariantQuery <-- Assembler : :query_config
VariantQuery --> Article : with :user, :query_config
GetRequest <-- Article : Article::ActiveRecord::Relation
@enduml
```
</details>

## Related Tickets & Documents

- Closes forem/forem#17272
- Closes forem/forem#17276
- Closes forem/forem#17216

## QA Instructions, Screenshots, Recordings

This code change brings us up to the point where we can flip the switch to use
the newly configured feed variant.

To test you will need to have some fresh articles (within the last 10 days).  Further, you will need to configure your test environment to note use the Basic feed strategy (Admin > Customization > User Exerpience and Brand > Feed strategy).  Last you'll need to enable the :feed_uses_variant_query_feature feature flag.  Then, while logged in, goto the `/` page and marvel at the feed.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/models/articles/feeds.rb [hunk 1]
```diff
@@ -84,19 +123,22 @@ def self.lever_catalog
 
       relevancy_lever(:comments_count,
                       label: "Weight to give to the number of comments on the article.",
+                      range: "[0..∞)",
```
Comment (joshpuetz) [inline app/models/articles/feeds.rb:126]: 😲

**Review by joshpuetz — APPROVED** [review]

File: app/models/articles/feeds.rb (hunk 1 above)
Comment (jeremyf) [reply app/models/articles/feeds.rb:126]: It's probably not infinity but it's definitely more than a dozen

File: app/models/articles/feeds/relevancy_lever.rb [hunk 2]
```diff
@@ -5,29 +5,96 @@ module Feeds
     #
     # @see config/feed/README.md
     class RelevancyLever
+      class ConfigurationError < StandardError
+      end
+
+      class InvaidFallbackError < ConfigurationError
```
Comment (maestromac) [inline app/models/articles/feeds/relevancy_lever.rb:11]: Typo here. 
```suggestion
      class InvalidFallbackError < ConfigurationError
```

File: app/models/articles/feeds/relevancy_lever.rb [hunk 3]
```diff
@@ -5,29 +5,96 @@ module Feeds
...
 
       alias user_required? user_required
+
+      # Responsible for configuring the lever with the given input.
+      #
+      # @param cases [Array<Array<Integer, Float>>]
+      # @param fallback [Float]
+      #
+      # @return [Articles::Feeds::RelevancyLever::Configured]
+      # @raise [Articles::Feeds::RelevancyLever::InvaidFallbackError] when the given fallback is
```
Comment (maestromac) [inline app/models/articles/feeds/relevancy_lever.rb:69]:
```suggestion
      # @raise [Articles::Feeds::RelevancyLever::InvalidFallbackError] when the given fallback is
```

File: app/models/articles/feeds/relevancy_lever.rb [hunk 4]
```diff
@@ -5,29 +5,96 @@ module Feeds
...
+      #        invalid.
+      # @raise [Articles::Feeds::RelevancyLever::InvaidCasesError] when the given cases is invalid.
+      def configure_with(cases:, fallback:)
+        raise InvaidFallbackError.new(fallback: fallback, key: key) unless valid_fallback?(fallback)
```
Comment (maestromac) [inline app/models/articles/feeds/relevancy_lever.rb:73]:
```suggestion
        raise InvalidFallbackError.new(fallback: fallback, key: key) unless valid_fallback?(fallback)
```

File: spec/models/articles/feeds/relevancy_lever_spec.rb [hunk 5]
```diff
@@ -13,8 +16,33 @@
...
+  describe "#configure_with" do
+    subject { lever.configure_with(fallback: fallback, cases: cases) }
+
+    let(:cases) { [[0, 0]] }
+    let(:fallback) { 1 }
+
+    context "when fallback is not a number" do
+      let(:fallback) { "nope" }
+
+      it { within_block_is_expected.to raise_error described_class::InvaidFallbackError }
```
Comment (maestromac) [inline spec/models/articles/feeds/relevancy_lever_spec.rb:33]:
```suggestion
      it { within_block_is_expected.to raise_error described_class::InvalidFallbackError }
```

**Pull request merged**